### PR TITLE
[NO SQUASH PLS] PlayerSAO::setHP - Don't call on_hpchange callbacks if HP hasn't changed

### PR DESCRIPTION
**Pull request by ClobberXD**
Also document the fact that `on_punchplayer` is invoked, even if the punched player is dead.

See #8462, and https://github.com/minetest/minetest/pull/8462#issuecomment-482242096

### CODE REVIEW

File: src/content_sao.cpp [hunk 1]
```diff
@@ -1336,11 +1336,13 @@ void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason)
 
 	hp = rangelim(hp, 0, m_prop.hp_max);
 
-	s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp - oldhp, reason);
-	if (hp_change == 0)
-		return;
+	if (m_hp == hp) {
```
Comment (rubenwardy) [inline src/content_sao.cpp]: shouldn't this be != ?

Comment (ClobberXD) [reply src/content_sao.cpp]: You didn't see that :D

**Comment by paramat** [comment]
I seem to remember there is a related issue, if so can you link to the issue?

**Comment by ClobberXD** [comment]
Yes, there is. Added link to first post.

File: src/content_sao.cpp [hunk 2]
```diff
@@ -1336,11 +1336,13 @@ void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason)
 
 	hp = rangelim(hp, 0, m_prop.hp_max);
 
-	s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp - oldhp, reason);
-	if (hp_change == 0)
-		return;
+	if (m_hp != hp) {
```
Comment (sfan5) [inline src/content_sao.cpp]: why does the `rangelim` code use oldhp, but this one uses m_hp?

Comment (ClobberXD) [reply src/content_sao.cpp]: There's no real reason for it - they're both the same anyway. I could make this check against `oldhp` instead if you want. I could also move the oldhp declaration below this if block to eliminate this confusion.

Comment (sfan5) [reply src/content_sao.cpp]: For consistency I'd use `oldhp` in both cases

Comment (ClobberXD) [reply src/content_sao.cpp]: Alright, done. :)

**Pull request merged**